File: src/back-end/FunChess/src/FunChess.Core.Chess/Structs/Move.cs
using System.Diagnostics.CodeAnalysis;
using System.Text.Json.Serialization;

namespace FunChess.Core.Chess.Structs;

public readonly struct Move
{
    public Move(Position previous, Position next)
    {
        Previous = previous;
        Next = next;
        DiffX = next.X - previous.X;
        DiffY = next.Y - previous.Y;
    }

    public Position Previous { get; }

    public Position Next { get; }

    public int DiffX { get; }

    public int DiffY { get; }

    public override string ToString()
    {
        return $"{Previous}|{Next}";
    }

    public static bool TryParse(string text,[NotNullWhen(true)] out Move? move)
    {
        move = null;
        
        ReadOnlySpan<char> span = text;
        if (span.IsEmpty) return false;
        
        int index = 1;
        while (index < span.Length && span[index] != '|')
        {
            ++index;
        }
        if (index + 1 == span.Length) return false;
        if (!int.TryParse(span[..index], out int previous)) return false;
        if (!int.TryParse(span.Slice(index + 1, span.Length - index - 1), out int next)) return false;

        try
        {
            Position previousPosition = new(previous);
            Position nextPosition = new(next);

            move = new Move(previousPosition, nextPosition);
            return true;
        }
        catch
        {
            return false;
        }
    }
}

File: src/front-end/fun-chess-app/src/core/chess/structs/Move.ts
import Position from "./Position";

export default class Move {
    public constructor(previous: Position, next: Position) {
        this.previous = previous;
        this.next = next;
        this.diffX = next.x - previous.x;
        this.diffY = next.y - previous.y;
    }

    public readonly previous: Position;

    public readonly next: Position;

    public readonly diffX: number;

    public readonly diffY: number;

    public toString(): string {
        return `${this.previous.index}|${this.next.index}`;
    }

    public static parse(text: string): Move {
        const indexes = text.split('|');
        if (indexes.length !== 2) throw Error("Text is not a valid move.");

        const previous = new Position(parseInt(indexes[0]));
        const next = new Position(parseInt(indexes[1]));

        return new Move(previous, next);
    }
}